source/_posts/2017-07-22-robot-framework-as-embedded-systems-validation-framework.markdown

```diff
@@ -0,0 +1,350 @@
+but limitation led me to search for different solution.
+`ser2net` and `telnet` solution was suggested on [mailing list](https://groups.google.com/d/msg/robotframework-users/r0xvLtGNgno/TI0suLOlNL4J)
+and eventually was good choice.
+
+```
+sudo apt-get install ser2net
+```
+
+Quick test with config file `ser2net_apu.cfg`
+
+```
+13542:telnet:600:/dev/ttyUSB0:115200 8DATABITS NONE 1STOPBIT
+```
+
+To prove that `ser2net` works correctly:
+
+```
+[0:19:39] pietrushnic:pcengines $ telnet localhost 13542
+Trying ::1...
+Connected to localhost.
+Escape character is '^]'.
+PC Engines apu2
+coreboot build 07/18/2017
+BIOS version
+4080 MB ECC DRAM
+
+
+telnet> q
+Connection closed.
+```
+
+### Telnet module for Robot Framework
+
+After playing some time I get to point when I can enter iPXE command prompt. My
+test looks pretty simple:
+
+```robot
+*** settings ***
+Library    Telnet
+
+*** Test Cases ***
+Enter PXE with 'n'
+    Open Connection    localhost    port=13542
+    Set Encoding    errors=strict
+    Set Timeout    30
+    Read Until    N for PXE boot
+    Write Bare    n
+    Read Until    Booting from ROM
+    Read Until    autoboot
+    # move arrow up
+    Write Bare    \x1b[A
+    Read Until    autoboot
+    Write Bare    \n
+    Read Until    iPXE>
+```
+
+Most complex part was related to pushing arrow keys through terminal. Magic is
+in `\x1b[A` what triggers escape sequence matching arrow up key on keyboard. Other keys are:
+
+```
+\x1b[B - down key
+\x1b[C - right key
+\x1b[D - left key
+```
+
+## Debugging pxelinux booting
+
+Serial console handling in Robot Framework is not trivial task. Especially, if
+you are doing it first time. What I learned is that below parameters are critical
+to correct understand what is going on behind the scene:
+
+* enable debug log by using `-b <file>` parameter
+* set debug level `-L <level>`
+
+Command for running framework should look like that:
+
+```
+robot -b debug.log -L TRACE <script_name>
+```
+
+### iPXE on apu2
+
+I tried to create Robot Framework script, but faced weird issue when trying to
+send more then one character. I described my findings in [this email](https://groups.google.com/d/msg/robotframework-users/5Mf2rKns13s/XQbalZ_DAQAJ).
+As I wrote it happen that `Telnet.write()` functions is too fast and iPXE
+cannot handle incoming characters. It took couple hours to figure out and
+without debugging output it would not be possible.
+
+## Booting pxelinux on apu2 using Robot Framework
+
+Finally I managed to boot to iPXE shell and reliably send commands. Next step
+was to provide address of PXE server for downloading and booting purpose.
+
+### Setup PXE server for apu2
+
+Communication with apu2 goes only through serial console or ssh when service
+ready, because of that typical Debian netboot had to be modified. In addition
+to that based on netboot package I decided to create bigger booting menu for
```

Comment:
"In addition to that I decided to create bigger booting menu, based on netboot package, for various systems..."